File: justtakethestairs/submissions/accepted/dp.py

N, C = map(int, input().split())
S = [0] * (N+1) #1-indexed
F = [0] * (N+1) #1-indexed
for i in range(1, N+1):
    S[i], F[i] = map(int, input().split())

solution = [float("inf")]*(N+1)
solution[0] = 0 #Base case
for i in range(1, N+1):
    students = 0
    floors = set()
    for j in range(i - 1, -1, -1):
        students += S[j+1]
        if students > C:
            break
        floors.add(F[j+1])
        max_floor = max(floors)
        stops = len(floors)
        trip_cost = max_floor + stops + max_floor
        if i == N:
            trip_cost -= max_floor #Elevator doesn't go down on last trip
        if solution[j] + trip_cost < solution[i]:
            solution[i] = solution[j] + trip_cost
print(solution[N])

File: justtakethestairs/submissions/accepted/dp.rs
use std::io::{self, BufRead, BufReader};
use std::collections::HashSet;
fn main() {
    let stdin = io::stdin();
    let mut lines = BufReader::new(stdin.lock()).lines();

    let line = lines.next().unwrap().unwrap();
    let mut iter = line.split_whitespace();
    let N: usize = iter.next().unwrap().parse().unwrap();
    let C: i64 = iter.next().unwrap().parse().unwrap();
    let mut S = vec![0i64; N+1];
    let mut F = vec![0i64; N+1];
    for i in 1..=N {
        let line = lines.next().unwrap().unwrap();
        let mut iter = line.split_whitespace();
        S[i] = iter.next().unwrap().parse().unwrap();
        F[i] = iter.next().unwrap().parse().unwrap();
    }

    let mut solution = vec![1_000_000_000i64; N+1]; //Same problem as kotlin solution
    solution[0] = 0;

    for i in 1..=N {
        let mut students: i64 = 0;
        let mut floors: HashSet<i64> = HashSet::new();
        for j in (0..i).rev() {
            students += S[j+1];
            if students > C {
                break;
            }
            floors.insert(F[j+1]);
            let max_floor = *floors.iter().max().unwrap();
            let stops = floors.len() as i64;
            let mut trip_cost = max_floor + stops + max_floor;
            if i == N {
                trip_cost -= max_floor;
            }
            if solution[j] + trip_cost < solution[i] {
                solution[i] = solution[j] + trip_cost
            }
        }
    }
    print!("{}", solution[N])
}